mlflow/types/schema.py

```diff
@@ -408,3 +408,165 @@ def __eq__(self, other) -> bool:
+        return {
+            "name": self.name,
+            "type": self.type,
+            "default": self.default,
+            "optional": self.optional,
+        }
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, ParamSpec):
+            return (
+                self.name == other.name
+                and self.type == other.type
+                and self.default == other.default
+                and self.optional == other.optional
+            )
+        return False
+
+    def __repr__(self) -> str:
+        default = f" (default: {self.default})" if self.default is not None else ""
+        optional = " (optional)" if self.optional else ""
+        return f"{self.name!r}: {self.type!r}{default}{optional}"
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `name` and `type` keys.
+        """
+        if not {"name", "type"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in ParamSpec JSON. Expected to find keys `name` and `type`"
+            )
+        return cls(
+            name=str(kwargs["name"]),
+            type=str(kwargs["type"]),
+            default=kwargs.get("default"),
+            optional=kwargs.get("optional", True),
+        )
+
+
+class ParamSchema:
+    """
+    Specification of parameters applicable to the model.
+    ParamSchema is represented as a list of :py:class:`ParamSpec`.
+    """
+
+    def __init__(self, params: List[ParamSpec]):
+        if not all(isinstance(x, ParamSpec) for x in params):
+            raise MlflowException(f"ParamSchema inputs only accept {ParamSchema.__class__}")
+        if self._has_duplicates(params):
+            raise MlflowException(
+                f"Duplicated parameters found in schema: {self._find_duplicates(params)}"
+            )
+        self._params = params
+
+    @staticmethod
+    def _has_duplicates(params: List[ParamSpec]) -> bool:
+        param_names = [param_spec.name for param_spec in params]
+        return len(param_names) != len(set(param_names))
+
+    @staticmethod
+    def _find_duplicates(params: List[ParamSpec]) -> List[str]:
+        param_names = [param_spec.name for param_spec in params]
+        uniq_param = set()
+        duplicates = []
+        for name in param_names:
+            if name in uniq_param:
+                duplicates.append(name)
+            else:
+                uniq_param.add(name)
+        return duplicates
+
+    def __len__(self):
+        return len(self._params)
+
+    def __iter__(self):
+        return iter(self._params)
+
+    @property
+    def params(self) -> List[ParamSpec]:
+        """Representation of ParamSchema as a list of ParamSpec."""
+        return self._params
+
+    def to_json(self) -> str:
+        """Serialize into json string."""
+        return json.dumps([x.to_dict() for x in self.params])
```

Comment:
```suggestion
        return json.dumps(self.to_dict())
```

Can we call `to_dict` here?